lib/puppet/resource_api.rb

```diff
@@ -430,6 +430,7 @@ def strict_check_title_parameter(current_state)
       end
 
       def context
+        self.class.context.catalog = @catalog if type_definition.feature?('raw_catalog_access')
         self.class.context
       end
```

Comment:
```suggestion
      def context
        @context ||= PuppetContext.new(TypeDefinition.new(definition), type_definition.feature?('raw_catalog_access') ? @catalog : nil)
      end
```

By creating the context for each instance separately, we can avoid the write access to `context.catalog` (see below).